### PR TITLE
solveLP() infeasible results fix

### LINKED ISSUE #12343: solveLP() produces infeasible results
##### System information (version)
- OpenCV => 3.0
- Operating System / Platform => Ubuntu 16.04 64-bit
- Compiler => gcc 4.9.4

##### Detailed description

I observed that `solveLP()` might produce infeasible results. When I do `A*x` Some of the constraints are not satisfied, i.e. `A*x > b` for some rows of `A`. Note that in my example `solveLP()` returns `SOLVELP_MULTI`, but it shouldn't matter it should still produce a feasible solution.

##### Steps to reproduce

I have attached the matrices to reproduce the issue. I have renamed them as argument names in the `solveLP()` signature. Just pass `Func` and `Constr` to `solveLP()` and obtain `z`. I have also attached `z` and `A*z` (`Constr_prod_z`)  for convenience.

You can observe that element at #161 of `Constr_prod_z` is `4.376743` which is clearly greater than `1` when it shouldn't be according to constraints (you can just print `Constr.row(161)` and look at the last element to see `1`). This is not the only example of constraint violation.

[mats.zip](https://github.com/opencv/opencv/files/2335467/mats.zip)

### PR DESCRIPTION
**Pull request by rayonnant14**
fix #12343 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by alalek** [comment]
Regression test case:
- should fail without the patch
- should pass with enabled patch.

Provided test case doesn't fail without the patch.

**Comment by rayonnant14** [comment]
@alalek thanks for comment. 
I made a change in the body of the test, it should be correct now

File: modules/core/test/test_lpsolver.cpp [hunk 1]
```diff
@@ -151,4 +151,20 @@ TEST(Core_LPSolver, issue_12337)
...
+TEST(Core_LPSolver, issue_12343)
+{
+    Mat A = (cv::Mat_<double>(4, 1) << 3., 3., 3., 4.);
+    Mat B = (cv::Mat_<double>(4, 5) << 0., 1., 4., 4., 3.,
+                                       3., 1., 2., 2., 3.,
+                                       4., 4., 0., 1., 4.,
+                                       4., 0., 4., 1., 4.);
+
+    Mat z;
+    int result = cv::solveLP(A,B,z);
```
Comment (alalek) [inline modules/core/test/test_lpsolver.cpp:163]: I believe, we still need some normalization or auto-tuning of `SOLVELP_EPS` value.

- `cv::solveLP(A,B,z);` => `OK`
- `cv::solveLP(A,B*2,z);` => `OK`
- `cv::solveLP(A,B*10,z);` => `OK`
- `cv::solveLP(A,B*1e3,z);` => `OK`
- `cv::solveLP(A,B*1e4,z);` => `FAIL`

`A` can be multiplied by any positive scalar too.

---

BTW, use `2**k` multiplier to minimize calculation errors in normalization (if normalization is going to be used).

File: modules/core/src/lpsolver.cpp [hunk 2]
```diff
@@ -259,10 +268,12 @@ static int inner_simplex(Mat_<double>& c, Mat_<double>& b,double& v,vector<int>&
         int e=-1,pos_ctr=0,min_var=INT_MAX;
```
Comment (alalek) [inline modules/core/src/lpsolver.cpp:268]: Please remove buggy "static" modifier from variable too

File: modules/core/src/lpsolver.cpp [hunk 3]
```diff
@@ -139,6 +139,15 @@ int solveLP(const Mat& Func, const Mat& Constr, Mat& z){
             *it=b.at<double>(indexToRow[i]-nsize,b.cols-1);
         }
     }
+    //check constraints feasibility
+    Mat prod = Constr(Rect(0, 0, Constr.cols - 1, Constr.rows)) * z;
+    Mat constr_check = (prod - 1e-12) > Constr.col(Constr.cols - 1);
```
Comment (alalek) [inline modules/core/src/lpsolver.cpp:144]:
```
Mat constr_check = Constr.col(Constr.cols - 1) - prod;
double minValue = 0;
minMaxIdx(constr_check, &minValue)  // we need minValue only
Analyze `minValue`
```

File: modules/core/include/opencv2/core/optim.hpp [hunk 4]
```diff
@@ -256,6 +256,7 @@ class CV_EXPORTS ConjGradSolver : public MinProblemSolver
 //! return codes for cv::solveLP() function
 enum SolveLPResult
 {
+    SOLVELP_LOST   = -3, //!< problem is feasible, but solver lost solution due to floating-point numbers representation error
```
Comment (alalek) [inline modules/core/include/opencv2/core/optim.hpp:259]: "due to floating-point arithmetic errors"

**Review by alalek — COMMENTED** [review]
Thank you for update!

File: modules/core/test/test_lpsolver.cpp [hunk 5]
```diff
@@ -158,13 +160,12 @@ TEST(Core_LPSolver, issue_12343)
...
-    Mat z;
-    int result = cv::solveLP(A,B,z);
-    EXPECT_EQ(result, 0);
-
-    Mat etalon_z = (cv::Mat_<double>(4, 1) << 16./37., 15./37., 0.0, 24./37.);
-    ASSERT_LT(cvtest::norm(z, etalon_z, cv::NORM_L1), 1e-12);
+    for (int i = 1; i < 1e5; i*=5)
+    {
+        Mat z;
+        int result = cv::solveLP(A,B*i,z);
```
Comment (alalek) [inline modules/core/test/test_lpsolver.cpp:166]: In linear task both A/B should be multiplied to keep the same result.

**Comment by asmorkalov** [comment]
@rayonnant14 Friendly reminder.

**Comment by asmorkalov** [comment]
@rayonnant14 Friendly reminder.

**Comment by asmorkalov** [comment]
@rayonnant14 Friednly reminder.

**Comment by asmorkalov** [comment]
@rayonnant14 Friednly reminder.

**Review by vpisarev — APPROVED** [review]

File: modules/core/test/test_lpsolver.cpp (hunk 5 above)
Comment (asmorkalov) [reply modules/core/test/test_lpsolver.cpp:166]: No it's not true. `5x+10y < 25` is the same as `x+2y < 5`. So mathematically it's correct test, but the idea of test is not obvious. Is the provided matrix is bed defined for Simplex Method?

File: modules/core/src/lpsolver.cpp [hunk 6]
```diff
@@ -285,11 +300,16 @@ static int inner_simplex(Mat_<double>& c, Mat_<double>& b,double& v,vector<int>&
         int row_it=0;
         MatIterator_<double> min_row_ptr=b.begin();
         for(MatIterator_<double> it=b.begin();it!=b.end();it+=b.cols,row_it++){
-            double myite=0;
-            //check constraints, select the tightest one, reinforcing Bland's rule
-            if((myite=it[e])>0){
+            if (fabs(it[b.cols-1]) <= SOLVELP_EPS_REL * fabs(it[b.cols-1]) + SOLVELP_EPS_ABS)
```
Comment (asmorkalov) [inline modules/core/src/lpsolver.cpp:303]: The check is equivalent to `(1-SOLVELP_EPS_REL)*fabs(it[b.cols-1]` <= SOLVELP_EPS_ABS` or `fabs(it[b.cols-1] < SOLVELP_EPS_ABS/0.2`. I do not think that it's expected.

**Pull request closed without merging**